File: Sources/CodexUsageWidget/CodexDataLocations.swift
import Foundation

struct CodexDataLocations: Equatable {
    let codexHomePath: String

    var stateDatabaseCandidates: [String] {
        [
            path("state_5.sqlite"),
            path("sqlite/state_5.sqlite")
        ]
    }

    var automationsDirectory: URL {
        URL(fileURLWithPath: path("automations"), isDirectory: true)
    }

    static func current(
        environment: [String: String] = ProcessInfo.processInfo.environment,
        homeDirectory: String = NSHomeDirectory()
    ) -> CodexDataLocations {
        if let override = environment["CODEX_HOME"]?.trimmingCharacters(in: .whitespacesAndNewlines),
           !override.isEmpty {
            return CodexDataLocations(codexHomePath: override)
        }

        return CodexDataLocations(codexHomePath: path(homeDirectory, ".codex"))
    }

    private func path(_ relativePath: String) -> String {
        Self.path(codexHomePath, relativePath)
    }

    private static func path(_ base: String, _ relativePath: String) -> String {
        var url = URL(fileURLWithPath: base, isDirectory: true)
        for component in relativePath.split(separator: "/") {
            url.appendPathComponent(String(component))
        }
        return url.path
    }
}

File: Platforms/Windows/CodexU.Windows/Services/CodexDataLocations.cs
namespace CodexU.Windows.Services;

public sealed class CodexDataLocations
{
    public CodexDataLocations(string codexHomePath)
    {
        CodexHomePath = codexHomePath;
    }

    public string CodexHomePath { get; }

    public IReadOnlyList<string> StateDatabaseCandidates => new[]
    {
        Path.Combine(CodexHomePath, "state_5.sqlite"),
        Path.Combine(CodexHomePath, "sqlite", "state_5.sqlite")
    };

    public string SessionsDirectory => Path.Combine(CodexHomePath, "sessions");

    public string ArchivedSessionsDirectory => Path.Combine(CodexHomePath, "archived_sessions");

    public string AutomationsDirectory => Path.Combine(CodexHomePath, "automations");

    public static CodexDataLocations Current()
    {
        var overrideHome = Environment.GetEnvironmentVariable("CODEX_HOME")?.Trim();
        if (!string.IsNullOrWhiteSpace(overrideHome))
        {
            return new CodexDataLocations(Environment.ExpandEnvironmentVariables(overrideHome));
        }

        var profile = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        if (string.IsNullOrWhiteSpace(profile))
        {
            profile = Environment.GetEnvironmentVariable("USERPROFILE") ?? "";
        }

        return new CodexDataLocations(Path.Combine(profile, ".codex"));
    }
}
